src/semantic_csv/core.clj

```diff
@@ -149,7 +149,75 @@
 ;; <br/>
 ;; Next, let's observe that `:this` and `:that` point to strings, while they should really be pointing to
 ;; numeric values.
-;; This can be addressed with the following function:
+;; This can be addressed with the following functions:
+
+
+;; ## cast-rules
+
+;; cast rules to be used in sniff-data
+;; this map should have a structure as such
+;; {:cast-class {:hierarch [:cast-type-1 :cast-type-2 ...]
+;;               :cast-type-1
+;;               :cast-type-2
+;;               ...}}
+;; The cast class defines the class of values (i.e. numeric values or date/time values).
+;; The cast type defines the semantic type within that class (i.e. integer, decimal, rational)
+(def cast-rules
+  {:numeric {:hierarchy [:integer :decimal :rational]
+             :integer impl/str->long
+             :decimal impl/str->double
+             :rational impl/str->rational}})
+
+;; <br/>
+;; ## sniff-data
+
+(declare cast-with)
+(defn sniff-data
+  "Investigates the data type of each column the rows and returns a vector of vectors.
+  Each vector points to a function within a map of functions that can be used for casting.
+  The 1-arity version of this function uses `cast-rules` for this. An optional map can be
+  passed to define the following options:
+
+  * `:do-cast` - Applies the cast-rules to each row using the `cast-with` function.
+  * `:rows-to-sniff` - An integer defining the number of rows used in sniffing for data types.
+  * `:cast-rules-map` - Map defining rules for sniffing and casting.
+  * `:cast-with-options` - Options to be passed to `cast-with`."
+  ([rows] (sniff-data {} rows))
+  ([{:keys [do-cast rows-to-sniff cast-rules-map cast-with-opts]
```

Comment:
A few thoughts/questions:
* I see what you're doing with `do-cast` here, making it possible to get back just the abstract result of the cast, as apposed to the casted data itself. However, the reason you would do one versus the other seems to be so different, that I wonder if the `do-cast` behavior shouldn't just be a separate function? Perhaps something like `autocast`? `sniff-cast` sounds a little silly, but could also be an option.
* Should `rows-to-sniff` be something `n-rows-to-sniff` or even just `n-rows`? Not convinced it's necessary, as short is somewhat nice, but a) we should probably stick to convention if it exists, and b) since I'm _not_ sure precedent exists (skimming the code), we should make sure of the pattern we want to set for the future, and I welcome your thoughts on this.
* Again, `cast-rules-map` seems like maybe it's something better suited to our second `autocast` (or whatever) function, if we end up doing that. Either way, maybe we just call it `cast-rules`? I think that fits better with established pattern wrt `cast-fns` (see #28).

Reply:
I plead guilty, for all silly naming here. Moreover, it's been a few years, so I had to sit down and read some of this code to refresh myself on what it all does.
I'll reply in order of the above bullets:

- `do-cast` is a silly name. another options `apply?`. If memory serves me, I wanted a way for the user to not have to dig into impl to get out values from `sniff-test`. It's handy in the repl. I'm not sure breaking it out into another function adds much value, either. If we did do that, it would manifest as what amounts to an alias to `sniff-test`. However, I'm totally game to do that and remove the option. And the more that I think about it, it's 6 in one hand, half dozen in the other. If we break them out into the applied version, I like `autocast`. The `sniff-test` alias could just be `sniff`.
- for `rows-to-sniff` - I like `n-rows` for brevity and reusability.
- `:%s/cast-rules-map/cast-rules/g` - agreed.